presto-main/src/main/java/io/prestosql/operator/LookupJoinOperator.java

```diff
@@ -41,587 +44,685 @@
-                    operatorContext.newAggregateSystemMemoryContext()));
+            return lookupSourceProviderFuture;
         }
 
-        PartitioningSpillResult result = spiller.get().partitionAndSpill(page, spillMask);
-        spillInProgress = result.getSpillingFuture();
-        return result.getRetained();
-    }
-
-    public LocalPartitionGenerator getPartitionGenerator()
-    {
-        if (!partitionGenerator.isPresent()) {
-            partitionGenerator = Optional.of(new LocalPartitionGenerator(hashGenerator, lookupSourceFactory.partitions()));
+        private boolean tryFetchLookupSourceProvider()
+        {
+            if (lookupSourceProvider == null) {
+                if (!lookupSourceProviderFuture.isDone()) {
+                    return false;
+                }
+                lookupSourceProvider = requireNonNull(getDone(lookupSourceProviderFuture));
+                statisticsCounter.updateLookupSourcePositions(lookupSourceProvider.withLease(lookupSourceLease -> lookupSourceLease.getLookupSource().getJoinPositionCount()));
+            }
+            return true;
         }
-        return partitionGenerator.get();
-    }
 
-    @Override
-    public Page getOutput()
-    {
-        // TODO introduce explicit state (enum), like in HBO
+        private PartitioningSpillResult spillAndMaskSpilledPositions(Page page, IntPredicate spillMask)
+        {
+            if (!spiller.isPresent()) {
+                spiller = Optional.of(partitioningSpillerFactory.create(
+                        probeTypes,
+                        getPartitionGenerator(),
+                        spillContext.newLocalSpillContext(),
+                        memoryTrackingContext.newAggregateSystemMemoryContext()));
+            }
 
-        if (!spillInProgress.isDone()) {
-            /*
-             * We cannot produce output when there is some previous input spilling. This is because getOutput() may result in additional portion of input being spilled
-             * (when spilling state has changed in partitioned lookup source since last time) and spiller does not allow multiple concurrent spills.
-             */
-            return null;
+            return spiller.get().partitionAndSpill(page, spillMask);
         }
-        checkSuccess(spillInProgress, "spilling failed");
 
-        if (probe == null && pageBuilder.isEmpty() && !finishing) {
-            return null;
+        public LocalPartitionGenerator getPartitionGenerator()
+        {
+            if (!partitionGenerator.isPresent()) {
+                partitionGenerator = Optional.of(new LocalPartitionGenerator(hashGenerator, lookupSourceFactory.partitions()));
+            }
+            return partitionGenerator.get();
         }
 
-        if (!tryFetchLookupSourceProvider()) {
-            if (!finishing) {
-                return null;
+        private WorkProcessor.TransformationState<WorkProcessor<Page>> tryUnspillNext()
```

Comment:
this code this method originated from is quite complex. What if `tryUnspillNext` behaved as the old method, that is call `restoreProbe`? This way it would make it easier to be sure that nothing critical changed behind the scenes.